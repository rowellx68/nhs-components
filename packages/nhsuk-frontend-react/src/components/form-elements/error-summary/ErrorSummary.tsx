'use client';

import React, { useRef, useImperativeHandle, useEffect } from 'react';
import { Factory, factory } from '@/internal/factory/factory';
import { useIdWithPrefix } from '@/internal/hooks/use-id-with-prefix';
import { AsElementProps, ElementProps } from '@/types/shared';
import clsx from 'clsx';
import {
  Heading,
  HeadingProps,
} from '@/components/styles/typography/heading/Heading';
import {
  PolymorphicFactory,
  polymorphicFactory,
} from '@/internal/factory/polymorphic-factory';
import {
  ErrorSummaryContextProvider,
  useErrorSummaryContext,
} from './ErrorSummary.context';
import { List, ListProps } from '@/components/styles/typography/list/List';
import { Base } from '@/components/core/base/Base';
import initErrorSummary from '@/resources/error-summary/error-summary';

export type ErrorSummaryProps = ElementProps<'div', 'role' | 'tabindex'>;

type ErrorSummaryFactory = Factory<{
  props: ErrorSummaryProps;
  ref: HTMLDivElement;
  staticComponents: {
    Title: typeof ErrorSummaryTitle;
    Body: typeof ErrorSummaryBody;
    List: typeof ErrorSummaryList;
    ListItem: typeof ErrorSummaryListItem;
  };
}>;

const ErrorSummary = factory<ErrorSummaryFactory>(
  ({ className, ...props }, ref) => {
    useImperativeHandle(ref, () => internalRef.current as HTMLDivElement);
    const internalRef = useRef<HTMLDivElement>(null);

    const labelId = useIdWithPrefix('error-summary');

    useEffect(() => {
      if (!internalRef.current) {
        return;
      }

      const parent = internalRef.current.parentElement;

      if (!parent) {
        return;
      }

      initErrorSummary({ scope: parent as any });
    }, [internalRef]);

    return (
      <ErrorSummaryContextProvider value={{ labelId }}>
        <div
          className={clsx('nhsuk-error-summary', className)}
          role="alert"
          aria-labelledby={labelId}
          tabIndex={-1}
          ref={internalRef}
          {...props}
        />
      </ErrorSummaryContextProvider>
    );
  },
);

export type ErrorSummaryTitleProps = Omit<HeadingProps, 'id'>;

type ErrorSummaryTitleFactory = PolymorphicFactory<{
  props: ErrorSummaryTitleProps;
  defaultComponent: 'h2';
  defaultRef: HTMLHeadingElement;
}>;

const ErrorSummaryTitle = polymorphicFactory<ErrorSummaryTitleFactory>(
  ({ as: component = 'h2', className, ...props }, ref) => {
    const { labelId } = useErrorSummaryContext();
    return (
      <Heading
        as={component}
        className={clsx('nhsuk-error-summary__title', className)}
        ref={ref}
        {...props}
        id={labelId}
      />
    );
  },
);

export type ErrorSummaryBodyProps = ElementProps<'div'>;

const ErrorSummaryBody = ({ className, ...props }: ErrorSummaryBodyProps) => {
  return (
    <div className={clsx('nhsuk-error-summary__body', className)} {...props} />
  );
};

export type ErrorSummaryListProps = Omit<ListProps, 'role'>;

const ErrorSummaryList = ({ className, ...props }: ErrorSummaryListProps) => {
  return (
    <List
      className={clsx('nhsuk-error-summary__list', className)}
      role="list"
      {...props}
    />
  );
};

export type ErrorSummaryListItemProps = ElementProps<'a'>;

type ErrorSummaryListItemFactory = PolymorphicFactory<{
  props: ErrorSummaryListItemProps;
  defaultComponent: 'a';
  defaultRef: HTMLAnchorElement;
}>;

const ErrorSummaryListItem = polymorphicFactory<ErrorSummaryListItemFactory>(
  (
    {
      as: component = 'a',
      ...props
    }: ErrorSummaryListItemProps & AsElementProps,
    ref,
  ) => {
    return (
      <li>
        <Base as={component} ref={ref} {...props} />
      </li>
    );
  },
);

ErrorSummary.displayName = 'ErrorSummary';
ErrorSummaryTitle.displayName = 'ErrorSummary.Title';
ErrorSummaryBody.displayName = 'ErrorSummary.Body';
ErrorSummaryList.displayName = 'ErrorSummary.List';
ErrorSummaryListItem.displayName = 'ErrorSummary.ListItem';

ErrorSummary.Title = ErrorSummaryTitle;
ErrorSummary.Body = ErrorSummaryBody;
ErrorSummary.List = ErrorSummaryList;
ErrorSummary.ListItem = ErrorSummaryListItem;

export {
  ErrorSummary,
  ErrorSummaryTitle,
  ErrorSummaryBody,
  ErrorSummaryList,
  ErrorSummaryListItem,
};
