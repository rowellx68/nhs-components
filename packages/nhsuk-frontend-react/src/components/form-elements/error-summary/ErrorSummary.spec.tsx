import React from 'react';
import { it, expect } from 'vitest';
import { render } from '@testing-library/react';
import { ErrorSummary } from './ErrorSummary';

it('should render the error summary component', () => {
  const { container } = render(
    <ErrorSummary>
      <ErrorSummary.Title>There is a problem</ErrorSummary.Title>
      <ErrorSummary.Body>
        <p>Describe the errors and how to correct them</p>
        <ErrorSummary.List>
          <ErrorSummary.ListItem href="#">
            Date of birth must be in the past
          </ErrorSummary.ListItem>
        </ErrorSummary.List>
      </ErrorSummary.Body>
    </ErrorSummary>,
  );

  expect(
    container.querySelector('.nhsuk-error-summary__title'),
  ).toHaveTextContent('There is a problem');
  expect(
    container.querySelector('.nhsuk-error-summary__body p'),
  ).toHaveTextContent('Describe the errors and how to correct them');
  expect(
    container.querySelector('.nhsuk-error-summary__list a'),
  ).toHaveTextContent('Date of birth must be in the past');
  expect(container).toMatchSnapshot();
});
