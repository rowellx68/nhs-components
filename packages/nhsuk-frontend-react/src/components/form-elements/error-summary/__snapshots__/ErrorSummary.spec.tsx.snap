// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render the error summary component 1`] = `
<div>
  <div
    aria-labelledby="error-summary_:r0:"
    class="nhsuk-error-summary"
    role="alert"
    tabindex="-1"
  >
    <h2
      class="nhsuk-error-summary__title"
      id="error-summary_:r0:"
    >
      There is a problem
    </h2>
    <div
      class="nhsuk-error-summary__body"
    >
      <p>
        Describe the errors and how to correct them
      </p>
      <ul
        class="nhsuk-list nhsuk-error-summary__list"
        role="list"
      >
        <li>
          <a
            href="#"
          >
            Date of birth must be in the past
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
